### ISSUE TITLE
Improve support for "Network" shares on Windows

### ISSUE THREAD

**Issue by CodingBadly** [fman-community/fman#266]
Nothing happens when CTRL + P and type '\\networkfolder\'

**Comment by mherrmann** [comment]
At the moment, the only way to get to a network folder is to enter the full path into GoTo. Eg.: `\\machine\folder`. Just `\\machine` unfortunately does not work. It's also not currently possible to get a list of all available network locations.

The goal of this issue is now to improve all of the above things:
- When starting to type `\\` into GoTo, it should suggest available network locations
- When typing `\\machine`, it should list the folders on this machine.

**Comment by mherrmann** [comment]
[fman 0.9.8](https://fman.io/blog/network-shares-on-windows/) added a first implementation for network shares on Windows. You can now see a list of shares when you press `Alt+F1/F2` and select `Network...`.

The points of this issue however still remain:
 * When starting to type `\\` into GoTo, fman should suggest available network locations
 * When typing `\\machine`, it should list the folders on this machine.

**Comment by kek91** [comment]
> When starting to type \\\ into GoTo, fman should suggest available network locations

I think you should also consider adding an artifical delay here when someone types "\\\\" before scanning the network. It doesnt have to be long, but like 0.2-0.3 seconds, just so fman doesnt start uneccessary network scans every time someone type \\\ as part of something else.

Also be aware that for medium/big sized networks the scan can take a while. For instance, at my job we have 3-400 computers/printers showing up in the "Network" list.

Quick brain storming

- spawn many threads for the scan to make it go faster
- wait 0.2 sec before initializing scan 
- cancel scan if user proceeds to write something after \\\ (or only scan items related to string entered)
- optionally cancel scan after 20-30 network objects/shares has been found
- or dont start scan until at least 1 letter after \\\ has been entered, to filter the results more (and reduce a scan with 1000s of objects to only 10s)

**Comment by mherrmann** [comment]
> just so fman doesnt start uneccessary network scans every time someone type \\ as part of something else.

I was thinking of only starting the scan when the search query _begins with_ "\\". Because then it would be clear it's a network path, right?

Thank you for the suggestions. I'm curious, do you think you'll have a chance to try `Alt+F1` then `Network...` at your job? Would be interesting what the performance is like currently.

**Comment by kek91** [comment]
Yes that's true. But if the user already knows where he's going, starting a scan would be unnecessary, hence the 0.2s delay. Alternatively just kill the scan threads and spawn new (filtered) scans if user types more letters (indicating part of the server name).

Maybe I just overthink it, as long as the fman UI doesn't hang I guess it doesn't really matter.

Oh boy, I just did some testing here and there's a couple of challenges. Testing from a Windows computer so I use `net view` to find shares.

- First of, running command `net view` (or `net view /all`) only lists shares on my local computer.
- Network list in Explorer shows only my own shares + printers on network, not shares from every server. I remember it used to list every computer/server shares as well so not sure if it's a change in Windows 10,  GPO rules in our domain or anything else.
- To list all shares from every server in a domain, `net view` requires parameter `/domain:company.com`
- Windows 10 has disabled SMBv1 after the cryptolocker attack earlier in 2017 so I don't know what happens if the scan detects a smbv1 share. I guess it doesn't see it at all.
- I just did two scans on the network. Once as domain administrator, once as my own user account. Both failed after ~50 seconds, error 64 and error 53.

I'm also not sure how you're doing the scans. Pinging every host on the subnet and checking for shares, or using the native methods (like `net view` on Windows and samba-something on Linux). 
So many variables here, difficult issue :)

When running the network scan in fman it only lists the shares from my own personal computer, just like Windows Explorer (except it doesn't include printers obviously).

I can try some more later tonight with Linux at another corp network, but it only has about 10 active computers so won't help much in terms of performance tests.

Edit: typo

**Comment by mherrmann** [comment]
Interesting, thank you very much for the research insights Kim. (Really helps because I don't have a company network I could use to test.)

When you say "it only lists the shares from my own personal computer", do you mean it actually only shows shares that "physically" lie on your personal computer? Because then, the "Network view" is actually kind of useless ;)

Thanks!
M

**Comment by kek91** [comment]
Glad to be of assistance :)

No not physical drives, but **network shares** which is configured to be shared from the computer. Windows doesn't share anything by default (*), so it's actually just an empty folder. Useless describes it accurately hehe

I feel we need more testing though, maybe there are some network restrictions in our environment. I mentioned earlier I can try from my Linux computer tonight but I failed to notice that the command is only available for Windows so far :)

I'll give it a try home later though and see how it works there.

*) Windows does have some shares by default (called administrative shares), but they are hidden and requires (minimum) read access on remote computer to access. For instance, every physical drive on a computer is actually shared as `\\computer\x$` but you will be prompted for credentials when trying to access it (unless you're in a domain and your domain user already has access to that share).
